### ISSUE TITLE
Hello_world_example_problem

### ISSUE THREAD

**Issue by hongleiw** [uwsampa/grappa#234]
Hi, I'm new in distributed system. Our lab want to use grappa. SO I'm trying to run Grappa's hello_world.exe after I configure it successfully and make demo-hello_world, then I have problem when I run it:
<img width="484" alt="screen shot 2015-10-17 at 9 31 28 pm" src="https://cloud.githubusercontent.com/assets/9388660/10562571/25412c18-7517-11e5-94ad-0d4b67e19b10.png">
How can I fix it? I notice there are red warning says about CXX, but I already put it in the path and gcc is 4.9 version and I can compile cpp using g++ and can run it.
<img width="489" alt="screen shot 2015-10-17 at 9 52 51 pm" src="https://cloud.githubusercontent.com/assets/9388660/10562608/803072c6-7519-11e5-806d-93a5e2c736a5.png">
<img width="487" alt="screen shot 2015-10-17 at 10 04 16 pm" src="https://cloud.githubusercontent.com/assets/9388660/10562636/2e4f4b10-751b-11e5-863f-4e9d759659d1.png">

Thank you in advance

**Comment by hongleiw** [comment]
And another confused thing is Do I need to mpicc some .c file and then run grappa applications using mpirun command? 
Actually I have question that MPI need first use mpicc to compile and then we can use mpirun to run it.
So similar, we need to somehow useing mpicc to compile, right? But seems I did't do this. Could you please tell me how to do it? More general, how grappa work with MPI?

So many questions... Hope not bother you.

**Comment by bmyerz** [comment]
Thanks for providing the log.
What is the output of
`mpicc -v` and `mpic++ -v`?
MPI needs to have been built with the same version of gcc as `gcc -v` and `g++ -v`

**Comment by bmyerz** [comment]
Can you be more specific about "how does grappa work with MPI" ?
The network transfer for Grappa's messaging (e.g. Grappa::delegate::call) is implemented with MPI

**Comment by nelsonje** [comment]
On the MPI question: 

Grappa's build system autodetects the MPI command/flags/libraries needed to build MPI binaries based on the ```mpicc``` in your path. Generally this should just work, although there are configuration options if you need to do something special.

On the error:

It looks like you're using a version of GCC you built yourself, which means your binary was linked agains a libstdc++.so somewhere in your gcc installation directory. But when you run the binary with mpirun it's trying to use the libstdc++.so in your standard OS install location instead. You should be able to fix this by either
* adding your GCC install's ```lib``` directory to the ```LD_LIBRARY_PATH``` environment variable in your shell before running the binary, or
* passing mpirun option to set the ```LD_LIBRARY_PATH``` environment variable. This works differently depending on which version of MPI you're using; check your mpirun man page for more info.

**Comment by hongleiw** [comment]
Yes, It make sense. And I added the gcc/lib to LD_LIBRARY_PATH, it stills goes wrong with the exact same error
<img width="1275" alt="screen shot 2015-10-20 at 8 57 55 pm" src="https://cloud.githubusercontent.com/assets/9388660/10643624/a7ced766-77d8-11e5-9b5d-aa8046c5c928.png">
And when I try the 2nd way, I use --prefix /home/hongleiw/lib/gcc/lib when I run mpirun, but seems it cannot recognize this configure.
My configuration is following
My mpi version is 
<img width="1270" alt="screen shot 2015-10-20 at 8 59 45 pm" src="https://cloud.githubusercontent.com/assets/9388660/10643720/30a62cec-77d9-11e5-8720-bf8f54b60cf6.png">
I change 2 line in grappa/configure and give --CC --CXX manually, 
<img width="484" alt="screen shot 2015-10-21 at 9 53 12 am" src="https://cloud.githubusercontent.com/assets/9388660/10643802/a711b91e-77d9-11e5-806a-a0c0eda44fa0.png">
The error should have nothing to do with this change, right?

**Comment by nelsonje** [comment]
Which version of MPI are you using?

**Comment by hongleiw** [comment]
Yes, it's MPICH 3.1.4

**Comment by nelsonje** [comment]
Usually when I see errors like what you saw originally, it's because mpirun/mpiexec is not propagating LD_LIBRARY_PATH to the processes it spawns on the worker nodes, so "ldd foo.exe" and "mpirun ldd foo.exe" will show different things. This can usually be fixed by telling mpirun explicitly to propagate LD_LIBRARY_PATH, but the syntax is different for each MPI. 

For MPICH with mpiexec.hydra, I believe it would be something like -genvlist LD_LIBRARY_PATH. Setting HYDRA_ENV=all in the submitting shell should automatically propagate all environment variables, too.

I'm installing MPICH 3.1.4 to see if I can replicate this.

**Comment by nelsonje** [comment]
Also, I notice now that the configure snippet you included looks like it's missing a recent commit. Could you tell me what ```git rev-parse --short HEAD``` says for your clone of the Grappa repo?

**Comment by hongleiw** [comment]
We find that it because the order of libraries in one of the cmake configuration files is incorrect. We change set(CMAKE_C_IMPLICIT_LINK_DIRECTORIES in CMakeCCompiler.cmake

Thank you!
